[READY] Let ycmd check rustup configured path as the last means to finding rust src path

bstaletic opened the pull request:
In Valloric/YouCompleteMe#2379 there was talk about enabling ycmd to check rust src path provided by rustup.

This allows users to have rust completion with zero ycmd configuration.

Tests are not included because tests themselves should behave differently based on how rustup was used and I'm not sure how to test that.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/785)
<!-- Reviewable:end -->

Valloric commented:
A testing approach I'd take would be to mock out the troubling parts. For instance, you can mock out `utils.SafePopen` to ensure the code that's calling it behaves correctly. Similar for `FindExecutable`, if necessary. 

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KnhIdOmxV3nwnZZDDlq:b-rng1b1)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Before I push the tests, one question.
To where should I point the `SafePopen()` in test that should end with valid completion results?
Or should I change the scripts in `ci` folder and not mock anything in that case?

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KnhSep3cbWWMbi9FjLy:b-z74w7l)*
<!-- Sent from Reviewable.io -->

Valloric commented:
> To where should I point the SafePopen() in test that should end with valid completion results?

I don't think I understand the question.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KnlSjJogJWfXhe8UT2F:b-edcjlw)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
`rustc --print sysroot` will always have some value. In case of `rustup` on my system that's `/home/bstaletic/.rustup/toolchains/stable-x86_64-unknown-linux-gnu`. Something like this path is what is expected of `SafePopen()` to return.

Do we have something like that? Do we have a valid rust src on continous integration bots?

I just think this needs a test where we show that semantic completion is working without any setup required.

I hope I was clearer now.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KnmCwhodKntt0FGilSv:b-7e19xo)*
<!-- Sent from Reviewable.io -->

micbou commented:
What I would do is create the `ycmd/tests/rust/testdata/rustup-toolchain/lib/rustlib/src/rust/src/libstd` folder with an empty file (e.g. `path.rs`), wrap the `rustc` call in a function, mock that function to return the path `ycmd/tests/rust/testdata/rustup-toolchain`, and send a `debug_info` request to check that the value of `rust_src_path` is equal to `ycmd/tests/rust/testdata/rustup-toolchain/lib/rustlib/src/rust/src`.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[ycmd/completers/rust/rust_completer.py, line 142 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-Knn_2_3fhCmyuSGfI3B:-Knn_2_4O47ZchSUwrBL:b-5aieon) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L142)):*
> ```Python
> 
>     # Try to figure out the src path using rustup
>     if not FindExecutable( 'rustc' ):
> ```

I would assign the result of this function to a variable and then use this variable in the `SafePopen` call.

---

*[ycmd/completers/rust/rust_completer.py, line 143 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-KnpFZxCdqTKEW1iDyvT:-KnpFZxCdqTKEW1iDyvU:b-hyazvj) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L143)):*
> ```Python
>     # Try to figure out the src path using rustup
>     if not FindExecutable( 'rustc' ):
>         return None
> ```

Indentation should be two spaces.

---

*[ycmd/completers/rust/rust_completer.py, line 145 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-KnoBakLKYihoR7v-lGA:-KnoBakLKYihoR7v-lGB:bkkgckl) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L145)):*
> ```Python
>         return None
> 
>     rust_sysroot = ToUnicode( utils.SafePopen( [ 'rustc',
> ```

`SafePopen` won't work on Windows because of [the `CREATE_NO_WINDOW` flag](https://msdn.microsoft.com/en-us/library/windows/desktop/ms684863.aspx). `rustc` will not print anything to stdout if there is no console window. We should directly use `subprocess.Popen`.

---

*[ycmd/completers/rust/rust_completer.py, line 150 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-Knn_B-WOTnsXQOh9PL3:-Knn_B-WOTnsXQOh9PL4:b8dic3b) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L150)):*
> ```Python
>                                                stdout = subprocess.PIPE )
>                                    .communicate()[ 0 ][ : -1 ] )
>     rust_src_path = p.join( rust_sysroot, 'lib/rustlib/src/rust/src' )
> ```

This should be rewritten to `p.join( rust_sysroot, 'lib', 'rustlib', 'src', 'rust', 'src' )` to get a proper path on Windows (a path with only `\` separators).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KnoE0D0I4ywsRraw4lY:bpv0v8q)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Yup, pretty much what @micbou said. :)

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[ycmd/completers/rust/rust_completer.py, line 145 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-KnoBakLKYihoR7v-lGA:-KnqeKaDXkwbghMuONUo:b-9y37j) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L145)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

`SafePopen` won't work on Windows because of [the `CREATE_NO_WINDOW` flag](https://msdn.microsoft.com/en-us/library/windows/desktop/ms684863.aspx). `rustc` will not print anything to stdout if there is no console window. We should directly use `subprocess.Popen`.
</blockquote></details>

Can we wrap that somehow or fix `utils.SafePopen`? I'd like to fix this for good so we don't have to think about it again. 

---

*[ycmd/completers/rust/rust_completer.py, line 150 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-Knn_B-WOTnsXQOh9PL3:-KnqeXa3NUGD-WJl8rHS:b6zrrk4) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L150)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

This should be rewritten to `p.join( rust_sysroot, 'lib', 'rustlib', 'src', 'rust', 'src' )` to get a proper path on Windows (a path with only `\` separators).
</blockquote></details>

Good catch!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-Knqe7BkTz-foD1R7vfT:big0wc9)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
I must me doing something stupid, but I can't figure out what's wrong. The tests I added all say:
```
TypeError: <My new test name> takes 1 positional argument but 2 were given
```

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/rust/rust_completer.py, line 142 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-Knn_2_3fhCmyuSGfI3B:-KnyZq_Lwdb4NKJkEf_0:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L142)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

I would assign the result of this function to a variable and then use this variable in the `SafePopen` call.
</blockquote></details>

Done.

---

*[ycmd/completers/rust/rust_completer.py, line 143 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-KnpFZxCdqTKEW1iDyvT:-KnyZwT5DZ9IEEFRXRaK:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L143)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Indentation should be two spaces.
</blockquote></details>

Done.

---

*[ycmd/completers/rust/rust_completer.py, line 145 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-KnoBakLKYihoR7v-lGA:-Kny_5gfkvIsjygaB8pB:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L145)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Can we wrap that somehow or fix `utils.SafePopen`? I'd like to fix this for good so we don't have to think about it again. 
</blockquote></details>

Done.

---

*[ycmd/completers/rust/rust_completer.py, line 150 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-Knn_B-WOTnsXQOh9PL3:-Kny_FOj0pNSLcQxgwfG:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L150)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Good catch!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-Knyvo9N3OM0P4L2fj1P:b-srvape)*
<!-- Sent from Reviewable.io -->

micbou commented:
That's because you are patching the `FindExecutable` function and thus passing an additional parameter to the tests. If you don't need the parameter, just add `*args` to the parameter list.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[ycmd/completers/rust/rust_completer.py, line 145 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-KnoBakLKYihoR7v-lGA:-KoF9DIpmPaJP7da8bI3:b-xg97u5) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L145)):*
<details><summary><i>Previously, bstaletic wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Found out that it works on Windows when we also open a pipe to `stdin` and `stderr`:
```python
utils.SafePopen( [ 'rustc', '--print', 'sysroot' ], stdin_windows = subprocess.PIPE, stdout = subprocess.PIPE, stderr = subprocess.PIPE )
```
so we could do that.

---

*[ycmd/completers/rust/rust_completer.py, line 76 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFF6kde0OTwOEvkXbf:-KoFF6kde0OTwOEvkXbg:b33z8gn) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/completers/rust/rust_completer.py#L76)):*
> ```Python
>                                         'sysroot' ],
>                                       stdout = subprocess.PIPE )
>                               .communicate()[ 0 ][ : -1 ] )
> ```

I would use [`rstrip`](https://docs.python.org/2/library/string.html#string.rstrip) to remove the trailing `\n`.

---

*[ycmd/tests/rust/get_completions_test.py, line 151 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFCU9D5CI6kjCM77Es:-KoFCU9D5CI6kjCM77Et:bseop4v) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/tests/rust/get_completions_test.py#L151)):*
> ```Python
> @IsolatedYcmd()
> @patch( 'ycmd.completers.rust.rust_completer._GetRustSysroot',
>         return_value = '/non/existing/file/path' )
> ```

Why don't we use `PathToTestFile` here?

---

*[ycmd/tests/rust/get_completions_test.py, line 163 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFC_mqf6sG92OuNw4P:-KoFC_mrJSHBbwwrlwmS:b1ak36s) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/tests/rust/get_completions_test.py#L163)):*
> ```Python
> @IsolatedYcmd()
> @patch( 'ycmd.completers.rust.rust_completer._GetRustSysroot',
>         return_value = abspath( 'ycmd/tests/rust/testdata/rustup-toolchain' ) )
> ```

Same comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KoFC0l1hv5mW8RKcTdx:bepgyp8)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
I've added `*args` to argument list, but it seems like my tests fail after sending the `/ready` request. Rust completer ends up writing this to log:
```
ycmd.completers.rust.rust_completer: ERROR: HTTPConnectionPool(host='127.0.0.1', port=53413): Max retries exceeded with url: /ping (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x40e2490>: Failed to establish a new connection: [Errno 111] Connection refused',))
```
The log is then followed by a traceback.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[ycmd/completers/rust/rust_completer.py, line 76 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFF6kde0OTwOEvkXbf:-KoGTvln2rTQBW2XpvQO:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/completers/rust/rust_completer.py#L76)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

I would use [`rstrip`](https://docs.python.org/2/library/string.html#string.rstrip) to remove the trailing `\n`.
</blockquote></details>

Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 151 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFCU9D5CI6kjCM77Es:-KoGWnJzLBXnrWbTF_iW:b-sd92sc) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/tests/rust/get_completions_test.py#L151)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Why don't we use `PathToTestFile` here?
</blockquote></details>

My bad. Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 163 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFC_mqf6sG92OuNw4P:-KoGWo1vROeQtuwKZ5nD:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/tests/rust/get_completions_test.py#L163)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Same comment.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KoGkyWkJP7t4956ASrR:bvrdfm1)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[ycmd/tests/rust/get_completions_test.py, line 163 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFC_mqf6sG92OuNw4P:-Krvxw1ssUdRSQqSXWsl:b-55iun1) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/tests/rust/get_completions_test.py#L163)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Why no `PathToTestFile` here?

---

*[ycmd/tests/rust/get_completions_test.py, line 166 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoHfmICIN02L7KTMwKy:-KoHfmICIN02L7KTMwKz:bef9als) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/tests/rust/get_completions_test.py#L166)):*
> ```Python
> def GetCompletions_RustupPathHeuristics_test( app ):
>   response = app.get( '/ready', { 'subserver': 'rust' } )
>   response = app.get( '/debug_info', { 'subserver': 'rust' } )
> ```

See [the Rust `DebugInfo` test](https://github.com/Valloric/ycmd/blob/0a8f6336f34f5edbfed6d1f75a9d7abe16e68cad/ycmd/tests/rust/debug_info_test.py#L32-L53) for an example on how to send the `/debug_info` request and parse its response.

---

*[ycmd/tests/rust/get_completions_test.py, line 139 at r3](https://reviewable.io:443/reviews/valloric/ycmd/785#-Krvy0bnqboVKza5yxd5:-Krvy0bnqboVKza5yxd6:b3gchfe) ([raw file](https://github.com/valloric/ycmd/blob/850594724fbced3248d170ceda885446bac3b16e/ycmd/tests/rust/get_completions_test.py#L139)):*
> ```Python
> 
> @IsolatedYcmd()
> @patch( 'ycmd.utils.FindExecutable', return_value = None )
> ```

There are two issues in this test. First, `FindExecutable` must be mocked from where it's looked up (this is explained [here](https://docs.python.org/3/library/unittest.mock.html#where-to-patch)) so it should be:
```
@patch( 'ycmd.completers.rust.rust_completer.FindExecutable', return_value = None )
```
Second, we don't raise the `NON_EXISTING_RUST_SOURCES_PATH_MESSAGE` exception when `rustc` is not found (because the Rust sources path may exist) but we still raise the `ERROR_FROM_RACERD_MESSAGE` exception when no completions is found in that case. This test should be written like the `WhenStandardLibraryCompletionFails_MentionRustSrcPath` one above. I would also rename it to something like `WhenRustcNotFound_MentionRustSrcPath`.

---

*[ycmd/tests/rust/get_completions_test.py, line 152 at r3](https://reviewable.io:443/reviews/valloric/ycmd/785#-Krw2fbIyOVHOlNPAxK5:-Krw2fbIyOVHOlNPAxK6:b-dx5who) ([raw file](https://github.com/valloric/ycmd/blob/850594724fbced3248d170ceda885446bac3b16e/ycmd/tests/rust/get_completions_test.py#L152)):*
> ```Python
> @patch( 'ycmd.completers.rust.rust_completer._GetRustSysroot',
>         return_value = '/non/existing/file/path' )
> def GetCompletions_NonExistingPathFromRustup_test( app, *args ):
> ```

Same issue as above: we don't raise the `NON_EXISTING_RUST_SOURCES_PATH_MESSAGE` exception when `rustc` is not found. I think this case is already covered by the `WhenStandardLibraryCompletionFails` and `WhenNoCompletionsFound` tests.

---

*[ycmd/tests/rust/get_completions_test.py, line 169 at r3](https://reviewable.io:443/reviews/valloric/ycmd/785#-Krw6JUCo1WNBCZgId8O:-Krw6JUCo1WNBCZgId8P:b-ln4ce0) ([raw file](https://github.com/valloric/ycmd/blob/850594724fbced3248d170ceda885446bac3b16e/ycmd/tests/rust/get_completions_test.py#L169)):*
> ```Python
>   entry = { 'rust_src_path':
>             abspath( 'ycmd/tests/rust/testdata/rustup-toolchain/lib/' +
>                      'rustlib/src/rust/src' ) }
> ```

`PathToTestFile`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[ycmd/completers/rust/rust_completer.py, line 145 at r1](https://reviewable.io:443/reviews/valloric/ycmd/785#-KnoBakLKYihoR7v-lGA:-KrySLQDHCorLQJ4JEe4:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/88a421614f2e5a739e200c71e84229ae0adc55ee/ycmd/completers/rust/rust_completer.py#L145)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Found out that it works on Windows when we also open a pipe to `stdin` and `stderr`:
```python
utils.SafePopen( [ 'rustc', '--print', 'sysroot' ], stdin_windows = subprocess.PIPE, stdout = subprocess.PIPE, stderr = subprocess.PIPE )
```
so we could do that.
</blockquote></details>

Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 163 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFC_mqf6sG92OuNw4P:-Krwdt1W2Bh-J8mDSb5Y:b-z8mxmt) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/tests/rust/get_completions_test.py#L163)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Why no `PathToTestFile` here?
</blockquote></details>

No good reason, fixed.

---

*[ycmd/tests/rust/get_completions_test.py, line 166 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoHfmICIN02L7KTMwKy:-KryGxWq0251CyUdDlFx:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/tests/rust/get_completions_test.py#L166)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

See [the Rust `DebugInfo` test](https://github.com/Valloric/ycmd/blob/0a8f6336f34f5edbfed6d1f75a9d7abe16e68cad/ycmd/tests/rust/debug_info_test.py#L32-L53) for an example on how to send the `/debug_info` request and parse its response.
</blockquote></details>

Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 139 at r3](https://reviewable.io:443/reviews/valloric/ycmd/785#-Krvy0bnqboVKza5yxd5:-KryR3sWhGYQ8KpL2Mi7:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/850594724fbced3248d170ceda885446bac3b16e/ycmd/tests/rust/get_completions_test.py#L139)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

There are two issues in this test. First, `FindExecutable` must be mocked from where it's looked up (this is explained [here](https://docs.python.org/3/library/unittest.mock.html#where-to-patch)) so it should be:
```
@patch( 'ycmd.completers.rust.rust_completer.FindExecutable', return_value = None )
```
Second, we don't raise the `NON_EXISTING_RUST_SOURCES_PATH_MESSAGE` exception when `rustc` is not found (because the Rust sources path may exist) but we still raise the `ERROR_FROM_RACERD_MESSAGE` exception when no completions is found in that case. This test should be written like the `WhenStandardLibraryCompletionFails_MentionRustSrcPath` one above. I would also rename it to something like `WhenRustcNotFound_MentionRustSrcPath`.
</blockquote></details>

Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 152 at r3](https://reviewable.io:443/reviews/valloric/ycmd/785#-Krw2fbIyOVHOlNPAxK5:-Krz4tseWd58BzY6seBG:bhmu5qy) ([raw file](https://github.com/valloric/ycmd/blob/850594724fbced3248d170ceda885446bac3b16e/ycmd/tests/rust/get_completions_test.py#L152)):*
> I think this case is already covered by the WhenStandardLibraryCompletionFails and WhenNoCompletionsFound tests.

After taking another look at thte code, I think you're right, so I removed the test.

---

*[ycmd/tests/rust/get_completions_test.py, line 169 at r3](https://reviewable.io:443/reviews/valloric/ycmd/785#-Krw6JUCo1WNBCZgId8O:-KrweMl7aTPP1AwTHIDg:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/850594724fbced3248d170ceda885446bac3b16e/ycmd/tests/rust/get_completions_test.py#L169)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

`PathToTestFile`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[ycmd/tests/rust/get_completions_test.py, line 169 at r3](https://reviewable.io:443/reviews/valloric/ycmd/785#-Krw6JUCo1WNBCZgId8O:-Ks-OufF7-4ZmOeafJ68:b4dxua1) ([raw file](https://github.com/valloric/ycmd/blob/850594724fbced3248d170ceda885446bac3b16e/ycmd/tests/rust/get_completions_test.py#L169)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

`PathToTestFile` should be used like `os.path.join` to obtain `\` on Windows and `/` on other platforms as separators:
```python
PathToTestFile( 'rustup-toolchain', 'lib', 'rustlib', 'src', 'rust', 'src' )
```

---

*[ycmd/tests/rust/get_completions_test.py, line 142 at r4](https://reviewable.io:443/reviews/valloric/ycmd/785#-Ks-PqQNftDooVYWZVuJ:-Ks-PqQNftDooVYWZVuK:b-zhy3gv) ([raw file](https://github.com/valloric/ycmd/blob/25963cc07cff9c05aab4643ee18ef313f79e26d7/ycmd/tests/rust/get_completions_test.py#L142)):*
> ```Python
> def GetCompletions_WhenRustcNotFound_MentionRustSrcPath_test( app, *args ):
>   filepath = PathToTestFile( 'rustup-toolchain/lib/rustlib/'
>                              'src/rust/src/libstd/path.rs' )
> ```

I don't think we want to complete in the Rust standard library but rather in `test.rs`.

---

*[ycmd/tests/rust/get_completions_test.py, line 165 at r4](https://reviewable.io:443/reviews/valloric/ycmd/785#-Ks-OnX4IN8MdZQonISI:-Ks-OnX5IX1cLN0nuPLn:b-povln) ([raw file](https://github.com/valloric/ycmd/blob/25963cc07cff9c05aab4643ee18ef313f79e26d7/ycmd/tests/rust/get_completions_test.py#L165)):*
> ```Python
>   request_data = BuildRequest( filetype = 'rust' )
> 
>   assert_that( app.post_json( '/debug_request', request_data ).json,
> ```

`/debug_info` instead of `/debug_request`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: 2 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 169 at r3](https://reviewable.io:443/reviews/valloric/ycmd/785#-Krw6JUCo1WNBCZgId8O:-KsAIMw8dWqps6bl9jEu:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/850594724fbced3248d170ceda885446bac3b16e/ycmd/tests/rust/get_completions_test.py#L169)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

`PathToTestFile` should be used like `os.path.join` to obtain `\` on Windows and `/` on other platforms as separators:
```python
PathToTestFile( 'rustup-toolchain', 'lib', 'rustlib', 'src', 'rust', 'src' )
```

</blockquote></details>

Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 142 at r4](https://reviewable.io:443/reviews/valloric/ycmd/785#-Ks-PqQNftDooVYWZVuJ:-KsAIK3kRBp7atmlGY7V:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/25963cc07cff9c05aab4643ee18ef313f79e26d7/ycmd/tests/rust/get_completions_test.py#L142)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

I don't think we want to complete in the Rust standard library but rather in `test.rs`.
</blockquote></details>

Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 165 at r4](https://reviewable.io:443/reviews/valloric/ycmd/785#-Ks-OnX4IN8MdZQonISI:-KsAIMFM5YO4SEymNuZr:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/25963cc07cff9c05aab4643ee18ef313f79e26d7/ycmd/tests/rust/get_completions_test.py#L165)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

`/debug_info` instead of `/debug_request`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
I fixed the `RustupPathHeuristics` test, but I can't figure out what's wrong with the `WhenRustcNotFound_MentionRustSrcPath` test.

---

Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KsBHEt7uANvgL1nxAWP:b1hhc6t)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 1 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[ycmd/completers/rust/rust_completer.py, line 76 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFF6kde0OTwOEvkXbf:-KsiEt3SSTmPe5f8oKzs:bocxtjq) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/completers/rust/rust_completer.py#L76)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I don't see the change.

---

*[ycmd/tests/rust/get_completions_test.py, line 58 at r6](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsiFRNmKxJuDMyf2sDi:-KsiFRNmKxJuDMyf2sDj:b4sc02m) ([raw file](https://github.com/valloric/ycmd/blob/cf7a1dd0101a443a507acdf42439e2377e70a450/ycmd/tests/rust/get_completions_test.py#L58)):*
> ```Python
> 
> @SharedYcmd
> def GetCompletions_WhenStandardLibraryCompletionFails_MentionRustSrcPath_test(
> ```

`_GetRustSysroot` should probably be patched to return a non-existing path for this test and the one below otherwise they 
fail if the user installed the Rust sources on its machine with rustup.

---

*[ycmd/tests/rust/get_completions_test.py, line 151 at r6](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsiFA15xy9uL8k-1bub:-KsiFA15xy9uL8k-1buc:by1yrmf) ([raw file](https://github.com/valloric/ycmd/blob/cf7a1dd0101a443a507acdf42439e2377e70a450/ycmd/tests/rust/get_completions_test.py#L151)):*
> ```Python
>                                   column_num = 1 )
> 
>   response = app.get( '/completions',
> ```

You need `post_json` here, not `get`. This should fix the test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/completers/rust/rust_completer.py, line 76 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFF6kde0OTwOEvkXbf:-Ksis4iN-I4vpCMYKwT_:b-p7v1zb) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/completers/rust/rust_completer.py#L76)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

I don't see the change.
</blockquote></details>

My bad. Now it is done.

---

*[ycmd/tests/rust/get_completions_test.py, line 58 at r6](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsiFRNmKxJuDMyf2sDi:-KsitgDCkHuNK538Ko79:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/cf7a1dd0101a443a507acdf42439e2377e70a450/ycmd/tests/rust/get_completions_test.py#L58)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

`_GetRustSysroot` should probably be patched to return a non-existing path for this test and the one below otherwise they 
fail if the user installed the Rust sources on its machine with rustup.
</blockquote></details>

Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 151 at r6](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsiFA15xy9uL8k-1bub:-Ksiv7Jwlw1vCjO6SQ4m:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/cf7a1dd0101a443a507acdf42439e2377e70a450/ycmd/tests/rust/get_completions_test.py#L151)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

You need `post_json` here, not `get`. This should fix the test.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 2 files at r7.
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/rust/rust_completer.py, line 76 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFF6kde0OTwOEvkXbf:-KsivlORTjrzJuSEVMX5:bpff8rm) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/completers/rust/rust_completer.py#L76)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

My bad. Now it is done.
</blockquote></details>

Should be `rstrip` without any parameter (`\n` is a whitespace character).

---

*[ycmd/tests/rust/get_completions_test.py, line 80 at r7](https://reviewable.io:443/reviews/valloric/ycmd/785#-Ksix596IoJBgYW4H_p2:-Ksix596IoJBgYW4H_p3:bntdl8y) ([raw file](https://github.com/valloric/ycmd/blob/3e09cf5ed8c2086c0830063d7fcdad82e5a5f447/ycmd/tests/rust/get_completions_test.py#L80)):*
> ```Python
> 
> @SharedYcmd
> def GetCompletions_WhenNoCompletionsFound_MentionRustSrcPath_test( app ):
> ```

This test should be patched like `GetCompletions_WhenStandardLibraryCompletionFails_MentionRustSrcPath` too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 61 at r7](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsixeUeknOu1klYToBW:-KsixeUeknOu1klYToBX:b-wg7ski) ([raw file](https://github.com/valloric/ycmd/blob/3e09cf5ed8c2086c0830063d7fcdad82e5a5f447/ycmd/tests/rust/get_completions_test.py#L61)):*
> ```Python
>         return_value = PathToTestFile( '/non/existing/rust/src/path' ) )
> def GetCompletions_WhenStandardLibraryCompletionFails_MentionRustSrcPath_test(
>   app ):
> ```

Need to add `*args` because of the patch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[ycmd/completers/rust/rust_completer.py, line 76 at r2](https://reviewable.io:443/reviews/valloric/ycmd/785#-KoFF6kde0OTwOEvkXbf:-KsixxxNX8CC4S5W-4I5:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/79a895b0d55d41dfe07f331b2d3cba20af3dc4a6/ycmd/completers/rust/rust_completer.py#L76)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Should be `rstrip` without any parameter (`\n` is a whitespace character).
</blockquote></details>

Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 61 at r7](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsixeUeknOu1klYToBW:-KsixwemevoCBVU3KhuB:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/3e09cf5ed8c2086c0830063d7fcdad82e5a5f447/ycmd/tests/rust/get_completions_test.py#L61)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Need to add `*args` because of the patch.
</blockquote></details>

Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 80 at r7](https://reviewable.io:443/reviews/valloric/ycmd/785#-Ksix596IoJBgYW4H_p2:-Ksiy5HGdWZjvhFl7jt-:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/3e09cf5ed8c2086c0830063d7fcdad82e5a5f447/ycmd/tests/rust/get_completions_test.py#L80)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

This test should be patched like `GetCompletions_WhenStandardLibraryCompletionFails_MentionRustSrcPath` too.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 57 at r7](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsiyEwb9IZb0AYJu9aE:-KsiyEwb9IZb0AYJu9aF:b-sh05v2) ([raw file](https://github.com/valloric/ycmd/blob/3e09cf5ed8c2086c0830063d7fcdad82e5a5f447/ycmd/tests/rust/get_completions_test.py#L57)):*
> ```Python
> 
> 
> @SharedYcmd
> ```

The test should be isolated: `@IsolatedYcmd()`.

---

*[ycmd/tests/rust/get_completions_test.py, line 79 at r8](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsiyX2N2GAYt1qkTPUN:-KsiyX2N2GAYt1qkTPUO:bwcdts0) ([raw file](https://github.com/valloric/ycmd/blob/fcd62aa8dd2dc68772cd2fe3523aeb3887561407/ycmd/tests/rust/get_completions_test.py#L79)):*
> ```Python
> 
> 
> @SharedYcmd
> ```

`@IsolatedYcmd()`

---

*[ycmd/tests/rust/get_completions_test.py, line 82 at r8](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsiyULBEemoZNpd72LE:-KsiyULBEemoZNpd72LF:b5n0bpw) ([raw file](https://github.com/valloric/ycmd/blob/fcd62aa8dd2dc68772cd2fe3523aeb3887561407/ycmd/tests/rust/get_completions_test.py#L82)):*
> ```Python
> @patch( 'ycmd.completers.rust.rust_completer._GetRustSysroot',
>         return_value = PathToTestFile( '/non/existing/rust/src/path' ) )
> def GetCompletions_WhenNoCompletionsFound_MentionRustSrcPath_test( app ):
> ```

Missing `*args`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: 2 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 57 at r7](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsiyEwb9IZb0AYJu9aE:-Ksiz7kQCPuW1e6p3QLG:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/3e09cf5ed8c2086c0830063d7fcdad82e5a5f447/ycmd/tests/rust/get_completions_test.py#L57)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

The test should be isolated: `@IsolatedYcmd()`.
</blockquote></details>

Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 79 at r8](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsiyX2N2GAYt1qkTPUN:-Ksiz9Hw6D-UhkNFua6n:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/fcd62aa8dd2dc68772cd2fe3523aeb3887561407/ycmd/tests/rust/get_completions_test.py#L79)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

`@IsolatedYcmd()`
</blockquote></details>

Done.

---

*[ycmd/tests/rust/get_completions_test.py, line 82 at r8](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsiyULBEemoZNpd72LE:-KsizNJQrJ1juLgGqsLa:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/fcd62aa8dd2dc68772cd2fe3523aeb3887561407/ycmd/tests/rust/get_completions_test.py#L82)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Missing `*args`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

micbou commented:
We are almost there.

---

Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 61 at r9](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsjD3bjWif-hrIs3cLL:-KsjD3bjWif-hrIs3cLM:b-sgc5rs) ([raw file](https://github.com/valloric/ycmd/blob/04cfbe52cd5dd0744dbebdc59e288c0c06165575/ycmd/tests/rust/get_completions_test.py#L61)):*
> ```Python
>         return_value = PathToTestFile( '/non/existing/rust/src/path' ) )
> def GetCompletions_WhenStandardLibraryCompletionFails_MentionRustSrcPath_test(
>   app, *args ):
> ```

We need to wait for the racerd server to be ready when a test is isolated. So, we should add
```python
WaitUntilCompleterServerReady( app, 'rust' )
```
like in the `GetCompletions_NoCompletionsFound_ExistingRustSrcPath` test. This applies to the `GetCompletions_WhenNoCompletionsFound_MentionRustSrcPath` and `GetCompletions_WhenRustcNotFound_MentionRustSrcPath` tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KsjDwhA0fb_QwtEpolO:b-41pt31)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 61 at r9](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsjD3bjWif-hrIs3cLL:-KsjHzBdk4IOndXPK4m6:bfs7uxc) ([raw file](https://github.com/valloric/ycmd/blob/04cfbe52cd5dd0744dbebdc59e288c0c06165575/ycmd/tests/rust/get_completions_test.py#L61)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

You forgot to add `WaitUntilCompleterServerReady( app, 'rust' )` to this test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 61 at r9](https://reviewable.io:443/reviews/valloric/ycmd/785#-KsjD3bjWif-hrIs3cLL:-KsjLA7k5R3G5XFCZqcu:b4ik359) ([raw file](https://github.com/valloric/ycmd/blob/04cfbe52cd5dd0744dbebdc59e288c0c06165575/ycmd/tests/rust/get_completions_test.py#L61)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

You forgot to add `WaitUntilCompleterServerReady( app, 'rust' )` to this test.
</blockquote></details>

Of course I did... Should be fixed now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
@micbou Thanks for the help. This should be READY, if no one has any other changes to request.

---

Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KsjoYZ4q1jTUFRwtP67:bi1s9qu)*
<!-- Sent from Reviewable.io -->

micbou commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KsjpG4IXOMlWtdcdZb_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

puremourning commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with one minor comment

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 59 at r11](https://reviewable.io:443/reviews/valloric/ycmd/785#-KuBXqbVPZJrqsBbczSR:-KuBXqbVPZJrqsBbczSS:b-oms4dn) ([raw file](https://github.com/valloric/ycmd/blob/7b40de9bd062cbc832c5321fc4940c6a6a87ec2d/ycmd/tests/rust/get_completions_test.py#L59)):*
> ```Python
> @IsolatedYcmd()
> @patch( 'ycmd.completers.rust.rust_completer._GetRustSysroot',
>         return_value = PathToTestFile( '/non/existing/rust/src/path' ) )
> ```

strictly we should use `PathToTestFile( 'non', 'existin', 'rust', 'src', 'path' )` so that the path gets correctly appended on Windows.

though i guess it doesn't really matter in _this_ case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KuBYGDOLRzpu-yqBZHp:b-7yu41d)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Reviewed 1 of 3 files at r2, 1 of 2 files at r8, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 59 at r11](https://reviewable.io:443/reviews/valloric/ycmd/785#-KuBXqbVPZJrqsBbczSR:-KuBYrBlDgq5S560gHlX:b11qf6q) ([raw file](https://github.com/valloric/ycmd/blob/7b40de9bd062cbc832c5321fc4940c6a6a87ec2d/ycmd/tests/rust/get_completions_test.py#L59)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

strictly we should use `PathToTestFile( 'non', 'existin', 'rust', 'src', 'path' )` so that the path gets correctly appended on Windows.

though i guess it doesn't really matter in _this_ case.
</blockquote></details>

Heh, I though I fixed that everywhere.
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 59 at r11](https://reviewable.io:443/reviews/valloric/ycmd/785#-KuBXqbVPZJrqsBbczSR:-KuB_4NDsNIloqbTf5QG:b-hk91oc) ([raw file](https://github.com/valloric/ycmd/blob/7b40de9bd062cbc832c5321fc4940c6a6a87ec2d/ycmd/tests/rust/get_completions_test.py#L59)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Heh, I though I fixed that everywhere.
Done.
</blockquote></details>

Why don't we just set `return_value` to `/non/existing/rust/src/path`? We don't need the `PathToTestFile` function here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[ycmd/tests/rust/get_completions_test.py, line 59 at r11](https://reviewable.io:443/reviews/valloric/ycmd/785#-KuBXqbVPZJrqsBbczSR:-KuBackAx5zZ4tzFc1Fn:b-ygh286) ([raw file](https://github.com/valloric/ycmd/blob/7b40de9bd062cbc832c5321fc4940c6a6a87ec2d/ycmd/tests/rust/get_completions_test.py#L59)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Why don't we just set `return_value` to `/non/existing/rust/src/path`? We don't need the `PathToTestFile` function here.
</blockquote></details>

Alright, changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

puremourning commented:
still <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with one tiny comment

---

Reviewed 1 of 1 files at r13.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 82 at r13](https://reviewable.io:443/reviews/valloric/ycmd/785#-KuC0kvdNFQ8txMe86Q2:-KuC0kvdNFQ8txMe86Q3:b-2guby3) ([raw file](https://github.com/valloric/ycmd/blob/8316a860e57880ea0cb275821babcf43887da5ef/ycmd/tests/rust/get_completions_test.py#L82)):*
> ```Python
> @IsolatedYcmd()
> @patch( 'ycmd.completers.rust.rust_completer._GetRustSysroot',
>         return_value = PathToTestFile( '/non/existing/rust/src/path' ) )
> ```

same here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KuC0nLhstAW2i1uSTwp:b-aujzb8)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/tests/rust/get_completions_test.py, line 82 at r13](https://reviewable.io:443/reviews/valloric/ycmd/785#-KuC0kvdNFQ8txMe86Q2:-KuC5tnSlRWK1rKArjBu:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/8316a860e57880ea0cb275821babcf43887da5ef/ycmd/tests/rust/get_completions_test.py#L82)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

same here
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785)*
<!-- Sent from Reviewable.io -->

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

This looks good to go!

@zzbot r=micbou

---

Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/785#-:-KuCQ154mziaZ9DcKd76:b-v6hn91)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
What happened to appveyor?

@zzbot retry
